### PR TITLE
Change program dependencies to cadence location

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: https://github.com/onflow/cadence/issues/1684

This is now possible that account freezing is removed, and will be needed to remove double counting of imports.

### CODE REVIEW

**Review by pattyshack — CHANGES_REQUESTED** [review]
I think this breaks "*" import (loading vs not loading an unrelated program in the same address will result in identical cadence execution, but different spock).  Can you add a test to ensure that's not the case?

**Comment by janezpodhostnik** [comment]
Thank you for pointing that out. It was problematic.

I still need dependencies per location for the next step. So I changed to track dependencies by both address and location and split invalidation to handle contract deployment differently from contract additions and removals.

File: fvm/environment/derived_data_invalidator.go [hunk 1]
```diff
@@ -101,14 +111,34 @@ func (invalidator ProgramInvalidator) ShouldInvalidateEntry(
 	}
 
 	// invalidate all programs depending on any of the contracts that were
-	// updated.  A program has itself listed as a dependency, so that this
+	// updated. A program has itself listed as a dependency, so that this
 	// simpler.
-	for _, key := range invalidator.ContractUpdateKeys {
-		_, ok := program.Dependencies[key.Address]
+	for _, key := range invalidator.ContractUpdates.Updates {
+		loc := common.AddressLocation{
```
Comment (pattyshack) [inline fvm/environment/derived_data_invalidator.go:117]: recreating the address location object on each ShouldInvalidateEntry call is expensive.  maybe do some pre-processing.  similar below for common.Address

Comment (janezpodhostnik) [reply fvm/environment/derived_data_invalidator.go:117]: I will address this in a separate PR when I switch everything to Address Location https://github.com/onflow/cadence/pull/2382

File: fvm/environment/contract_updater.go [hunk 2]
```diff
@@ -437,26 +437,50 @@ func (updater *ContractUpdaterImpl) RemoveContract(
...
+		shouldDelete := len(v.Code) == 0
+
+		if shouldDelete {
+			// this is a removal
+			contractUpdates.Deletions = append(contractUpdates.Deletions, v.ContractUpdateKey)
+			err = updater.accounts.DeleteContract(v.Name, v.Address)
+			if err != nil {
+				return ContractUpdates{}, err
+			}
+		} else if currentlyExists {
```
Comment (pattyshack) [inline fvm/environment/contract_updater.go:466]: maybe combine the else if / else.  ContractExists don't need to be called when deleting.  SetContract can be call unconditionally

**Review by pattyshack — APPROVED** [review]

**Comment by janezpodhostnik** [comment]
bors merge

**Review by koko1123 — APPROVED** [review]

**Comment by janezpodhostnik** [comment]
bors merge

**Pull request merged**